crates/ruff_python_parser/src/semantic_errors.rs

```diff
@@ -1688,6 +1710,36 @@ impl Visitor<'_> for ReboundComprehensionVisitor<'_> {
     }
 }
 
+#[derive(Default)]
+struct ReturnVisitor {
+    return_range: Option<TextRange>,
+    has_yield: bool,
+}
+
+impl StatementVisitor<'_> for ReturnVisitor {
+    fn visit_stmt(&mut self, stmt: &Stmt) {
+        match stmt {
+            Stmt::Expr(ast::StmtExpr { value, .. }) => match **value {
+                Expr::Yield(_) | Expr::YieldFrom(_) => {
+                    self.has_yield = true;
+                }
+                _ => {}
+            },
+            Stmt::FunctionDef(_) | Stmt::ClassDef(_) => {}
+            Stmt::Return(ast::StmtReturn {
+                value: Some(_),
+                range,
+                ..
+            }) => {
+                if self.return_range.is_none() {
+                    self.return_range = Some(*range);
+                }
```

Comment:
I like that this would take the first return like CPython does rather than the last return like B901, but I think for the sake of compatibility with B901, we should just take the last return and always overwrite:


```suggestion
                self.return_range = Some(*range);
```